src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +182,108 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
   return bestFoundGain;
 }
 
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<double>& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::Row<double> sortedLabels(labels.n_elem);
+  arma::rowvec sortedWeights;
+  for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+    sortedLabels[i] = labels[sortedIndices[i]];
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Only initialize if we are using weights.
+  if (UseWeights)
+  {
+    sortedWeights.set_size(sortedLabels.n_elem);
+    // The weights must keep the same order as the labels.
+    for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+      sortedWeights[i] = weights[sortedIndices[i]];
+  }
+
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
+  // Force a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  // Loop through all possible split points, choosing the best one.
+  for (size_t index = minimum; index < data.n_elem - minimum + 1; ++index)
+  {
+    // Make sure that the value has changed.
+    if (data[sortedIndices[index]] == data[sortedIndices[index - 1]])
+      continue;
+
+    /* TODO: The following function calculates the gain for each split each time from scratch
+             This can be greatly improved using advanced techniques like prefix sum and
+             prefix sum of squares etc. This will have drastic effects on runtime and is
+             definitely something we would want in future.
+     */
+    // Calculate the gain for the left and right child.
+    const double leftGain = FitnessFunction::template Evaluate<UseWeights>(sortedLabels,
+        sortedWeights, 0, index);
+    const double rightGain = FitnessFunction::template Evaluate<UseWeights>(sortedLabels,
+        sortedWeights, index, labels.n_elem);
+
+    double gain = leftGain + rightGain;
```

Comment:
Actually I wonder if we should be using a weighted gain computation here.  Instead of taking `leftGain + rightGain` and using unnormalized MSE (e.g. squared error), use MSE as it was before your most recent commit, and then use `leftGain * (numPointsOnLeft / totalPoints) + rightGain * (numPointsOnRight / totalPoints)`. I believe this should give the same result, but it keeps the name "MSE" accurate. :)

Reply:
Sure, this sounds good. I will make this change 👍